python/ray/data/_internal/dataset_iterator/pipelined_dataset_iterator.py

```diff
@@ -35,11 +35,20 @@ def _to_block_iterator(
     ]:
         epoch_pipeline = self._get_next_dataset()
 
+        if epoch_pipeline._first_dataset is not None:
+            blocks_owned_by_consumer = (
+                epoch_pipeline._first_dataset._plan.execute()._owned_by_consumer
+            )
+        else:
+            blocks_owned_by_consumer = (
+                epoch_pipeline._peek()._plan.execute()._owned_by_consumer
+            )
```

Comment:
could you add a comment in code for why we need to do this?

Reply:
Done.